mlir-sycl/include/mlir/Dialect/SYCL/Analysis/AliasAnalysis.h

```diff
@@ -0,0 +1,47 @@
+//===- AliasAnalysis.h - SYCL Alias Analysis -------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file contains the implementation of a alias analysis for the SYCL
+// dialect.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef MLIR_DIALECT_SYCL_ANALYSIS_ALIASANALYSIS_H
+#define MLIR_DIALECT_SYCL_ANALYSIS_ALIASANALYSIS_H
+
+#include "mlir/Analysis/AliasAnalysis/LocalAliasAnalysis.h"
+
+namespace mlir {
+namespace sycl {
+
+/// Specialized alias analysis for SYCL dialect operations.
+class AliasAnalysis : public LocalAliasAnalysis {
+protected:
+  AliasResult aliasImpl(Value lhs, Value rhs) override;
+
+private:
+  /// Return 'NoAlias' if both values are function arguments and any of them
+  /// have attribute 'local_alias_analysis.restrict', and an empty optional
+  /// otherwise.
+  Optional<AliasResult> handleRestrictAlias(Value lhs, Value rhs);
```

Comment:
This is not SYCL related, and is actually not needed for the test case, do you think we should remove it?

Reply:
I know is not SYCL related but I do think is useful nevertheless. I would keep it unless there are strong objections.